Enable MCUX_PIT with multiple pit channels

EmilioCBen opened the pull request:
Enable MCUX_PIT with multiple pit channels

decsny commented on soc/arm/nxp_imx/rt/soc_rt11xx.c:
```diff
@@ -334,6 +334,11 @@ static ALWAYS_INLINE void clock_init(void)
+#if defined(CONFIG_COUNTER_MCUX_PIT)
+	rootCfg.mux = kCLOCK_BUS_LPSR_ClockRoot_MuxSysPll3Out;
+	rootCfg.div = 2;
+	CLOCK_SetRootClock(kCLOCK_Root_Bus_Lpsr, &rootCfg);
```
I am not sure it is appropriate to do this code based on whether CONFIG_COUNTER_MCUX_PIT is defined, since it changes the configuration of the LPSR root, which is for all of the low power peripherals in the chip

EmilioCBen replied:
I am open to an alternative,
At the moment, 
IF the device is an 1166_CM7 or an 1176_CM7
then this is the only spot where the LPSR is configured,
and is the first time LPSR gets configured which means it can be
overwritten by the more important peripheral clocking setups.
Another thing to consider is that the PIT2 can only be rooted from the LPSR Root.

decsny replied:
I think we just don't need to guard it with #if PIT

EmilioCBen replied:
Check this again, I put this in a more convenient spot.

decsny replied:
My comment was not about code format, it was a functional concern with conditional compilation paths

EmilioCBen replied:
The LPSR Clock root is no longer conditioned by the PIT Kconfig

decsny commented on drivers/counter/counter_mcux_pit.c:
```diff
@@ -179,41 +213,101 @@ static const struct counter_driver_api mcux_pit_driver_api = {
+		.base = (PIT_Type *)DT_INST_REG_ADDR(n),					\
+		.irq_config_func = mcux_pit_irq_config_func_##n,				\
+		.num_channels = DT_INST_PROP(n, num_channels),					\
+		.pit_period = DT_INST_PROP(n, pit_period),					\
```
this property and struct field (pit_period) needs to be removed

zephyr api says default counter top shall be max value, and this is not used anywhere

EmilioCBen replied:
I have removed the pit period field and will be updated the next push to this PR.

manuargue reviewed (changes requested):
This is breaking support for S32Z2 boards, see https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/arm/nxp/nxp_s32z27x_rtu0_r52.dtsi#L92. Please do the updates needed for these boards and their test overlays.

decsny commented on drivers/counter/counter_mcux_pit.c:
```diff
@@ -130,17 +151,24 @@ static uint32_t mcux_pit_get_frequency(const struct device *dev)
-	if (data->top_callback) {
-		data->top_callback(dev, data->top_user_data);
+
+	for (int channel_index = 0;
```
@manuargue it looks like in the init macro all the interrupts are being connected to this routine, and this routine checks the peripheral register for which channels interrupted, so what did you mean is the issue?

decsny replied:
@manuargue you are right, but previously this driver didn't even support multiple channels, so I would like to ask is it necessary for author to implement this right now to merge

decsny replied:
Okay, now I agree that this is not an optimization, and it should be done in this PR, because of the contract of the interrupt priorities in devicetree.

decsny replied:
@manuargue Okay, I agree now that this is not an optimization because it is required by the contract of interrupts in DT. So, if there is an SOC that has the pit interrupts like this, then it should be done this way. But, I don't see an SOC in the tree that has multiple interrupts like this. Do you know of an SOC that does? Theoretically this argument could be made for any driver in zephyr, and most drivers do not have instance-based ISRs

decsny replied:
Thank you I missed this, so, I agree this is currently a bug in the PR for the k6x series

EmilioCBen replied:
I am still working on this.

EmilioCBen replied:
Both case scenarios of the interrupts are handled now with the small caveat of have conditional #ifdefs due to the way we are currently connecting the array of IRQs with its corresponding channel data device.

decsny commented on dts/bindings/counter/nxp,mcux-pit-channel.yaml:
```diff
@@ -0,0 +1,13 @@
+description: |
+  Child node for the Periodic Interrupt Timer node, intended for an individual timer channel
+
+compatible: "nxp,mcux-pit-channel"
```
devicetree tooling does not generate non-generic property macros on nodes without a compatible binding

EmilioCBen replied:
The new binding will be required if the parent includes child bindings which seem to be 
more inline with the other nodes that include channels.
The channels need their own binding due to the register value being required on the parent node.

mmahadevan108 commented on dts/bindings/counter/nxp,mcux-pit.yaml:
```diff
@@ -3,9 +3,9 @@
 description: NXP MCUX Periodic Interrupt Timer (PIT)
 
-compatible: "nxp,kinetis-pit"
+compatible: "nxp,mcux-pit"
```
This would break backward compatibility. Please document in the migration guide.

decsny replied:
> This would break backward compatibility. Please document in the migration guide.

yes, and all the other changes in this PR break backward compatibility, so we should update the migration guide anyways @EmilioCBen

EmilioCBen replied:
> if we are going to rename this binding I'd suggest `nxp,pit`, because MCUX is making reference to an SDK, not hw related. And also to mention it in release notes so downstream users are aware of this change.

One thing I was considering when I changed the binding is that
IF the IMX boards and now with my awareness of the S32 boards support the pit 
THEN should this be considered a kinetis pit suggesting that only works for the kinetis.
However another thing to consider is that currently this is not bare metal and still
uses the APIs from the MCUX SDK so I changed it to mcux,pit.

Do you think I should still change it to nxp,pit?

EmilioCBen replied:
I will update the migration guide in the next push to this PR.

EmilioCBen replied:
I will also update the ST Boards

danieldegrasse replied:
> IMO it should be nxp,pit. The binding is a description of a device. There could be different driver implementations (baremtal, SDKs, etc) for the same binding.

I agree here. We (hopefully) won't end up with multiple different driver implementations, but there is no need to include the software HAL name in the binding compatible. Devicetree is fundamentally intended to describe hardware, and the hardware here is an instance of the PIT IP.

EmilioCBen replied:
I updated the binding to be nxp,pit
I have updated the s32 boards
I have updated the migration guide.

dleach02 replied:
@manuargue we don't have your boards so you will need to test for us.

EmilioCBen committed:
boards: arm: Updated the board docs for pit

Updated the board documentation to reflect
the enablement of the PIT peripheral.

Signed-off-by: Emilio Benavente <emilio.benavente@nxp.com>

EmilioCBen commented on dts/bindings/counter/nxp,mcux-pit.yaml:
```diff
@@ -14,10 +14,10 @@ properties:
     required: true
 
-  pit-channel:
+  num-channels:
```
Still looking into this

EmilioCBen replied:
Thanks for the suggestion, num-channels is no longer a required property

decsny commented on drivers/counter/counter_nxp_pit.c:
```diff
@@ -0,0 +1,407 @@
+	const struct nxp_pit_config *config = dev->config;
+	int channel_id = PIT_CHANNEL_ID(dev);
+
+	PIT_DisableInterrupts(config->base, channel_id,
```
what's the need to enable/disable interrupts in start/stop functions?

decsny replied:
also I guess the SDK handles the difference between when theres one interrupt vs many?

EmilioCBen replied:
At this point in the code flow, we are dealing with individual devices, a device per channel, I imagine if you have the ability, you would want to ensure interrupts are available as the timer is active and disabled when the timer is no longer of use to the user, there may be more potential to look at in regards to these interrupts but might be a subject for a different PR.

decsny commented on drivers/counter/counter_nxp_pit.c:
```diff
@@ -0,0 +1,407 @@
+
+	flags = PIT_GetStatusFlags(config->base, channel_id);
+
+	return ((flags & mask) == mask);
```
should return 1 or 0 not true or false

EmilioCBen replied:
? Those are equivalent statements, are they not? 
Even with regards to the way they are defined by the API
I am sure the one who documented the API understood
the returning statement 1 as "true, there are pending interrupts"
and 0 as "false, there are no pending interrupts"

decsny replied:
I guess I was taught and it was drilled into me that the relational operators and true are not guaranteed to be 1... but on some research to the C standard, it seems maybe this teaching was wrong or the C standard changed in later revisions... C11 says == will yieldand int 1 if true, so maybe it's fine then

decsny commented on drivers/counter/counter_nxp_pit.c:
```diff
@@ -0,0 +1,407 @@
+		channel_index < config->num_channels;
+		channel_index++) {
+		PIT_SetTimerPeriod(config->base, channel_index,
+			USEC_TO_COUNT(config->info.max_top_value, clock_rate));
```
nit: this code is the same between compilation paths, this source can be restructured to be simpler

EmilioCBen replied:
The code paths are not the same, in one code path a single interrupt config function is called, and in the other we ask the channel if they have an interrupt config function and if so we use their individual config functions, at the moment the driver will have either one case or the other.

decsny replied:
what I mean is something like this:
```
	for (int channel_index = 0;
		channel_index < config->num_channels;
		channel_index++) {
#if DT_NODE_HAS_PROP(DT_COMPAT_GET_ANY_STATUS_OKAY(nxp_pit), interrupts)
		if (config->irq_config_func[channel_index]) 
#endif
                {
			config->irq_config_func[channel_index](dev);
			PIT_SetTimerPeriod(config->base, channel_index,
				USEC_TO_COUNT(config->info.max_top_value, clock_rate));
		}
	}
```

but maybe check that compliance won't flag this

decsny replied:
okay, just forget about this suggestion, its really not important

decsny commented on drivers/counter/counter_nxp_pit.c:
```diff
@@ -0,0 +1,407 @@
+		},										\
+		.base = (PIT_Type *)DT_INST_REG_ADDR(n),					\
+		.irq_config_func = nxp_pit_irq_config_func_##n,					\
+		.num_channels = DT_INST_FOREACH_CHILD(n, NXP_PIT_INC_FOR_CHANNEL_COUNT) 0,	\
```
cant use sep macro?

decsny replied:
`DT_INST_FOREACH_CHILD_SEP_VARGS(n, DT_NODE_HAS_COMPAT, (+), nxp_pit_channel)`

manuargue reviewed (changes requested):
Copyright legends need updates

decsny commented on drivers/counter/counter_nxp_pit.c:
hi Manuel,

I am not the author of this PR, but I am trying to help it to get merged, 

I tend to agree about the multiple driver implementation, but I wonder if the nxp pit file can be made bare metal and we can just remove the mcux driver? @Emilio would this be scope creep? Because I think the peripheral is so simple enough, it might actually be more work to rename it again than to just convert these Hal calls to register writes, just saying

dleach02 replied:
@manuargue, lets not block this PR for a file name change. This PR has been outstanding for a long time and we need to get it merged in (ideally before the freeze at the end of the week). If in a future PR we want to have multiple drivers with the same binding then that PR can change the file names to align with what you desire.

EmilioCBen replied:
The bare metal version of the driver should be possible but I believe it should be the discussion of a different PR, as far as the diff is concerned I have added a commit for just the name change and another commit for the driver changes, I will update this shortly.

EmilioCBen commented on doc/releases/migration-guide-3.6.rst:
```diff
@@ -405,6 +428,8 @@ Other Subsystems
   zbus options are renamed. Instead, the new :kconfig:option:`ZBUS_MSG_SUBSCRIBER_BUF_ALLOC_DYNAMIC`
   and :kconfig:option:`ZBUS_MSG_SUBSCRIBER_BUF_ALLOC_STATIC` options should be used.
 
+* Add something about the Kconfig and driver name from mcux-pit to nxp-pit
```
haha I usually add marks that make check compliance catch this for me, but I guess I slipped up here.

EmilioCBen commented on boards/arm/mimxrt1010_evk/mimxrt1010_evk.dts:
```diff
@@ -1,6 +1,6 @@
 /*
  * Copyright (c) 2023 TiaC Systems
- * Copyright 2019,2023 NXP
+ * Copyright 2019-24 NXP
```
Fixed

EmilioCBen commented on boards/arm/mimxrt1060_evk/mimxrt1060_evk.dts:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2018, NXP
+ * Copyright 2018-24 NXP
```
Fixed

EmilioCBen commented on boards/arm/mimxrt1160_evk/mimxrt1160_evk_cm7.dts:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2021, NXP
+ * Copyright 2021-24 NXP
```
Fixed

EmilioCBen commented on boards/arm/mimxrt1170_evk/mimxrt1170_evk_cm7.dts:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright 2021-22, NXP
+ * Copyright 2021-24, NXP
```
Fixed

EmilioCBen commented on drivers/clock_control/clock_control_mcux_ccm.c:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2017, NXP
+ * Copyright 2017-24 NXP
```
Fixed

EmilioCBen commented on drivers/counter/Kconfig.nxp_pit:
```diff
@@ -0,0 +1,12 @@
+# Periodic Interrupt Timer (PIT)
+
+# Copyright 2020-24 NXP
```
Fixed

EmilioCBen commented on drivers/counter/counter_nxp_pit.c:
```diff
@@ -0,0 +1,407 @@
+/*
+ * Copyright 2020-24 NXP
```
Fixed

EmilioCBen commented on dts/arm/nxp/nxp_rt11xx.dtsi:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright 2021,2023 NXP
+ * Copyright 2021-24 NXP
```
Fixed

EmilioCBen commented on dts/arm/nxp/nxp_s32z27x_rtu0_r52.dtsi:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright 2022-2023 NXP
+ * Copyright 2022-24 NXP
```
Fixed

EmilioCBen commented on dts/arm/nxp/nxp_s32z27x_rtu1_r52.dtsi:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright 2022-2023 NXP
+ * Copyright 2022-24NXP
```
Fixed

EmilioCBen commented on dts/bindings/counter/nxp,pit.yaml:
```diff
@@ -0,0 +1,20 @@
+# Copyright 2020-24 NXP
```
Fixed

EmilioCBen commented on include/zephyr/dt-bindings/clock/imx_ccm.h:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2017-2022, NXP
+ * Copyright 2017-2024 NXP
```
Fixed

EmilioCBen commented on include/zephyr/dt-bindings/clock/imx_ccm_rev2.h:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2021, NXP
+ * Copyright 2021-24 NXP
```
Fixed

EmilioCBen commented on soc/arm/nxp_imx/rt/soc_rt11xx.c:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright 2021-2023 NXP
+ * Copyright 2021-24 NXP
```
Fixed

EmilioCBen commented on tests/drivers/counter/counter_basic_api/boards/mimxrt1060_evk.overlay:
```diff
@@ -1,3 +1,23 @@
+/*
+ * Copyright 2023-24 NXP
```
Fixed

manuargue reviewed (commented):
all my comments addressed, approved with this remark https://github.com/zephyrproject-rtos/zephyr/pull/66336#discussion_r1472214411 and outstanding nits pointed out by other reviewers

mmahadevan108 commented on drivers/counter/CMakeLists.txt:
```diff
@@ -33,7 +33,7 @@ zephyr_library_sources_ifdef(CONFIG_COUNTER_MCUX_LPTMR          counter_mcux_lpt
 zephyr_library_sources_ifdef(CONFIG_COUNTER_NATIVE_POSIX        counter_native_posix.c)
 zephyr_library_sources_ifdef(CONFIG_USERSPACE                   counter_handlers.c)
-zephyr_library_sources_ifdef(CONFIG_COUNTER_MCUX_PIT            counter_mcux_pit.c)
+zephyr_library_sources_ifdef(CONFIG_COUNTER_NXP_PIT             counter_nxp_pit.c)
```
This rename should be documented in the 3.6  migration guide.

EmilioCBen replied:
I have added some documentation about the Kconfig change in the migration guide.

mmahadevan108 commented on dts/arm/nxp/nxp_s32z27x_rtu0_r52.dtsi:
```diff
@@ -89,12 +89,50 @@
 
 		pit0: pit@76150000 {
-			compatible = "nxp,kinetis-pit";
+			compatible = "nxp,pit";
```
Document this change to the binding name in the 3.6 migration guide.

EmilioCBen replied:
I have documented the change, is there something more specific you want me to note in regards to the binding?

decsny commented on drivers/counter/counter_mcux_pit.c:
```diff
@@ -1,219 +0,0 @@
- * SPDX-License-Identifier: Apache-2.0
- */
-
-#define DT_DRV_COMPAT nxp_kinetis_pit
```
try adding a commit somewhere in here that just renames the file, and maybe that diff will be more useful than nothing for people reading the commit history in the future, you don't need to do the work of renaming everything in the code in the file, just please try making a commit for just renaming the file so there is at least some diff shown

also it would be good to rename the kconfigs in that commit too since it is somewhat user facing (that change is not too hard I think, just a find and replace) and update the migration guide in that commit as well, like @mmahadevan108  suggested

decsny replied:
manifest change also needs to be in the commit where the kconfig changes

decsny reviewed (changes requested):
I am wanting to get this PR merged before feature freeze but I am going to put this block until the comments are addressed

EmilioCBen commented on drivers/counter/counter_nxp_pit.c:
```diff
@@ -0,0 +1,407 @@
+
+#define NXP_PIT_INC_FOR_CHANNEL_COUNT(ignored) 1 +
+
+#if DT_NODE_HAS_PROP(DT_COMPAT_GET_ANY_STATUS_OKAY(nxp_pit), interrupts)
```
I would disagree on the macro becoming easier to read, here we have two init macros, one per compilation run and both kept in their own scope of the file, if I create only one init macro as you suggest, the two changes within the the init macro will be required to become submacros(config struct and the irq config definitions) as you say, I think this will look equally if not more complicated then the current design. The config struct as an example makes up a large portion of the Init macro, since one config contains an array of pointer where the other does not (.irq_config_func) this will have to become two sub macros one per compilation run. Secondly the irq macros that were specific to the second init macro now become specific to a unified init macros which makes less sense to someone new looking at the file.
If you still think the opposite I can make the change, just understand that the code will be equally difficult if not harder to read for newer readers of the file.

decsny replied:
I very much disagree but I'm not blocking over this since it might be personal preference

But one advantage to how I described is to have single point of control, since there is so much duplicated code between these two macros that is exactly the same, you wouldn't have to make changes in two places in the future

decsny replied:
resolved

mmahadevan108 commented:
@EmilioCBen, please update `west.yml` to pull the latest changes for this PR.

EmilioCBen commented:
> @EmilioCBen, please update `west.yml` to pull the latest changes for this PR.

Updated!

EmilioCBen committed:
dts: bindings: renamed binding used for pit

Renamed pit binding from nxp,kinetis-pit
to nxp,pit

Signed-off-by: Emilio Benavente <emilio.benavente@nxp.com>

EmilioCBen committed:
drivers: counter: update nxp_pit api names

Update the api named for the new nxp pit
driver.

Signed-off-by: Emilio Benavente <emilio.benavente@nxp.com>

decsny commented on dts/bindings/counter/nxp,pit.yaml:
```diff
@@ -1,11 +1,11 @@
-# Copyright 2020,2023 NXP
+# Copyright 2020,2024 NXP
```
needs to be 2020, 2023-2024

EmilioCBen replied:
Updated

danieldegrasse commented on boards/arm/mimxrt1010_evk/mimxrt1010_evk.dts:
Commit message for this commit: `dts: arm: nxp: adding multiple nxp,pit channels` should probably be updated to something like:

```
drivers: counter: counter_nxp_pit: add support for multiple channels

Add support for multiple channels to the PIT counter driver. Board and SOC
devicetrees are also updated to reflect the new multi-channel pit support
```

danieldegrasse replied:
Another suggestion here- you could likely split the changes in this commit, so that only boards using the PIT currently have their devicetrees updated in this PR. Then the next commit could enable PIT for new boards.

EmilioCBen replied:
Both comments have been addressed

danieldegrasse commented on doc/releases/migration-guide-3.6.rst:
```diff
@@ -279,6 +279,33 @@ Device Drivers and Device Tree
+  to :dtcompatible:`nxp,pit` and has been updated to support multiple channels.
+  To configure the individual channels, you must add a child node with the
+  compatible :dtcompatible:`nxp,pit-channel` and configure as below.
+  The :kconfig:option:`CONFIG_COUNTER_MCUX_PIT` has also been to
```
Typo here, I assume you meant to say "has also been renamed to"

EmilioCBen replied:
Thank you for noticing, updated!

decsny commented:
another minor thing is the migration guide info about the changes to the binding/kconfig should probably be in the same commit as the changes

EmilioCBen committed:
drivers: counter: Adding Mutli Channel PIT Support

Updating the nxp,pit driver to support mutliple
channels. Updating the dts and board overlays
to account for the changes.

Signed-off-by: Emilio Benavente <emilio.benavente@nxp.com>

EmilioCBen committed:
dts: arm: Enabling pit for imx devices

Enabled the PIT and Multi channel support
for some of the rtXXXX devices.
- rt1010
- rt1060
- rt1160
- rt1170

Signed-off-by: Emilio Benavente <emilio.benavente@nxp.com>

EmilioCBen commented:
> another minor thing is the migration guide info about the changes to the binding/kconfig should probably be in the same commit as the changes

Updated

nordicjm reviewed (dismissed):
Needs converting to hwmv2

EmilioCBen commented:
> Needs converting to hwmv2

Should be resolved now

decsny commented:
@henrikbrixandersen please revisit

decsny commented:
@nordic-krch can you please visit this PR, you're assignee

dleach02 commented:
Merging as @nordic-krch +1 was removed when we updated the manifest to point to the merged HAL side.
